text/0000-cargo-libstd-awareness.md

```diff
@@ -0,0 +1,159 @@
+Then, if packages can explicitly depend on core they can also specify whether the float feature is needed.
+Cargo would be able to infer both of these options by inspecting all crates in the dependency graph.
+
+Rustbuild must currently perform multiple `cargo builds`, the first to build the standard library and the rest to build things which depend on the standard library.
+If rustc, rustfmt, etc, and their deps (some of which come from crates.io, and thus aren't specially tailored for building rust) declare deps on std, rustbuild wouldn't need multiple lockfiles.
+Keeping multiple in sync is nuisance, and this gets us one step closer to a single `cargo build` building rustc.
+
+The use-cases so far mainly benefit niche corners of the Rust community, but the last should be useful for just about everyone.
+Now that multiple versions of Rust have been released, it can be useful to specify the minimum version.
+If and when Rust 2, a version with breaking changes, comes out, this will be all the more important.
+We don't yet have a plan yet to track the language itself, but by tracking standard library dependencies we make it trivial to specify version requirements like any other package.
+
+
+# Detailed design
+
+The subsections with "interface" in their title form the normitive part of this RFC.
+The rest of this section just illustrate hows they would likely be implemented.
+
+## `Cargo.toml` Interface
+
+First and foremost, one will now be able to depend on standard library crates by version, e.g. `std = "1.10"`.
+This will work just as if `std` was on crates.io---features and other modifiers are supported.
+
+For backwards compatibility, Cargo must inject such standard library dependencies for existing packages.
+Exactly which dependencies is unresolved, but a requirement at least as strong as `std = "^1.0"` as a primary and build dependency is assured.
+
+Now, not all crates depend on `std`, so there must be a way to opt out.
+For this, we introduce a new `implicit-dependencies` key.
+It is defined by default as:
+```toml
+implicit-dependencies = ["primary", "build", "dev"]
+```
+This indicates each of `dependencies`, `build-dependencies`, and `dev-dependencies` maps (respectively) is augmented with implicit elements.
+A manual definition may be that or almost any subset, in which case only the included dependency maps are augmented.
+The one additional rule is `"dev"` must be included in the set: we have no plan for Cargoizing the default test runner (either the attribute syntax manipulation or runtime), and wish to be forward-compatible with doing so in the future.
+
+Finally, if an (explicit) dependency conflicts with one of the implicit defaults, that category of implicit dependency will be skipped.
+For example, if a crate explicit depends on `std` as a build-dependency, neither `std` nor any other implicit build dependency will be injected.
```

Comment:
My idea was rather than error when an explicit dep conflicts with an implicit dep, just drop the implicit dep. Since `core` would probably be an implicit dep, the only package which would need to use the `implicit-dependencies` key in practice would be core itself. Everything else would either use implicit deps, have an explicit dep on core, or have an explicit dep on std.

This is just convenient short hand. For simplicity's sake I am completely fine going back to making it an error as it would naturally be.